### PR TITLE
meta: adopt next-intl and app router

### PR DESCRIPTION
**Pull request by araujogui**
## Description

This PR refactors the Node.js Website and adopts the new Next.js App Router. This PR does all the necessary changes to make the app router-compatible. Bear in mind that some changes here are temporary.

**Note:** This PR is still going to have some updates, such as improvement of documentation, updates on Collaborator Guides, etc.

### Changes

- Removed a bunch of React Providers of our "static data" (such as blog data) and make them available directly on Hooks
- Removed `react-intl` and our custom i18n middleware in favour of `next-intl`
  - [Adoption of `next-intl` v3](https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/)
- Refactored our i18n keys to follow the ICU format that `next-intl` uses
- Updated several components to use `next-intl` and its components
- Removed `LocalizedLink`, `useLocale`, etc in favor of `next-intl` builtin's
- Refactored `next.dynamic.mjs`, our App Root Layout, and all the core `app/` folders to make the whole process of building and rendering compatible.
- Other minor changes that are related to make this compatible
- Removed HTMLHead and added App Router Metadata generation
- Moved Node Release Data Parsing directly to Release Data Generation
- Removed unused types, utilities and Hooks
- Tried to simplify and optimize existing Layouts and removed duplicated Layouts
- Overall cleanup of Application logic
- Updated Unit Tests to make them work
- Fixed imports

### Caveat

This rework is **stable**, but `next-intl` 3.0 is still on Release Candidate, and a few things _might_ change. For example, the usage of an unstable API https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/docs/getting-started/app-router#add-unstable_setrequestlocale-to-all-layouts-and-pages

## Validation

Builds should pass, Linting should pass, Deployment should pass, all pages should work as expected.

Deep testing is needed here!

### CODE REVIEW

**Comment by canerakdas** [comment]
I took a quick look;

- In Home Page, internal links are not working as expected
- Header Node.js logo redirect to home page won't work as expected

**Comment by ovflowd** [comment]
> I took a quick look;
> 
> * In Home Page, internal links are not working as expected
> * Header Node.js logo redirect to home page won't work as expected

Unable to reproduce locally, just on the preview. If I land in any other pages all the links work, except if I'm on the Home page

**Review by canerakdas — COMMENTED** [review]
## Localized content

Docs page sidebar navigation tags don't match. `$LTS`, `$ARCHIVED`

[Preview](https://nodejs-8uo2jpcub-openjs.vercel.app/en/docs)
image

[Nodejs.org
](https://nodejs.org/en/docs)
image

## Date Format

Wrong date format. Should be Day Month Year in my locale

### Blog Detail

[Preview](https://nodejs-8uo2jpcub-openjs.vercel.app/en/blog/vulnerability/openssl-fixes-in-regular-releases-oct2023)
image

[Nodejs.org
](https://nodejs.org/en/blog/vulnerability/openssl-fixes-in-regular-releases-oct2023)image

### Blog Listing Page

[Preview](https://nodejs-8uo2jpcub-openjs.vercel.app/en/blog)
image

[Nodejs.org
](https://nodejs.org/en/blog)
image

**Comment by ovflowd** [comment]
The $ARCHIVED and $LTS is just a little typo on DocsLayout.

About DateFormat the formatting is happening on the server so I need to check how to get the user time format, otherwise we shall simply keep the server format imo

**Comment by ovflowd** [comment]
FYI @canerakdas the time and date formatting will follow the current applied locale (https://next-intl-docs-git-feat-next-13-rsc-next-intl.vercel.app/docs/usage/dates-times#custom-date-and-time-formats)

This is better, so it always follows the locale correctly both on server-side and local-side

**Comment by canerakdas** [comment]
Thank you for the updates @ovflowd, I will read the doc 👍 

During my checks, I noticed that the URLs in the sitemap were not correct in the [preview](https://nodejs-25pg6shul-openjs.vercel.app/sitemap.xml), I think we need to look here too 🕵️‍♂️

```diff
+  <loc>https://nodejs-25pg6shul-openjs.vercel.app/about/governance</loc>
-  <loc>https://nodejs-25pg6shul-openjs.vercel.app/en/about/governance</loc>
```

**Comment by ovflowd** [comment]
> Thank you for the updates @ovflowd, I will read the doc 👍
> 
> During my checks, I noticed that the URLs in the sitemap were not correct in the [preview](https://nodejs-25pg6shul-openjs.vercel.app/sitemap.xml), I think we need to look here too 🕵️‍♂️
> 
> ```diff
> +  <loc>https://nodejs-25pg6shul-openjs.vercel.app/about/governance</loc>
> -  <loc>https://nodejs-25pg6shul-openjs.vercel.app/en/about/governance</loc>
> ```

Good catch. Fixed, forgot to add locales on routes :)

Thanks for doing this QA process, @canerakdas we probably want to test these paths in the future.

**Comment by ovflowd** [comment]
Hey @nodejs/web-infra and @nodejs/nodejs-website, the PR is finally now ready. I've done all the latest changes and fixes, cleanups I've could think of.

I know this is a big PR but would love reviews 🙇

**Review by AugustinMauroy — APPROVED** [review]
LGTM ! _so clean infra_
I'm still not fan of having one big catch all segment.

**Comment by ovflowd** [comment]
> I'm still not fan of having one big catch all segment.

It's the only way to support our approach of internationalisation with subfolders dynamically, and without creating specific catch-all segments. It's also the most perfomatic way, otherwise we would need to also calculate other subsets of directories and do other things.

I tried to make sub-folder catch-all segments to each subfolder (about, learn, etc) and have a specific layout for each of them, but there were numerous issues that raised. Hence having just one catch-all segment.

Performance-wise, the way we're doing Layouts has no cost compared if we did dedicated Next.js layouts, not to mention, for the legacy website, we have different root layouts do the structure of the website.

Maybe with the redesign, we could re-eval how this works, as all pages pretty much follow an unified design.

So there's room for just one root layout, that renders a left sidebar and a right sidebar conditionally... (But that's something we need to investigate, and how the layout would know what sort of metadata to load, and etc...)

In other words, abstraction is possible, but complex. In this case, I feel that having one sort of layout for each sort of section of the website that uses different sort of metadata/components, might be more "manual" but it's less costly and easier to understand.

But let's see, as I said, I'll try to address these topics once we start evaluating the layout work...

cc @bmuenzenmeyer if you have 2 cents here

**Comment by ovflowd** [comment]
Hey y'all, as I believe that a full review of this PR is pretty hard, and to unblock further development. I'm queueing this to our merge queue.

Please feel free to do any sort of reviews post-merge. I've battle-tested this and so far all good, so let's do it.

**Comment by ovflowd** [comment]
cc @nodejs/web-infra regarding my message above ☝️

File: app/[locale]/[[...path]]/page.tsx [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+type DynamicParams = { params: DynamicStaticPaths };
+
+// This is the default Viewport Metadata
+// @see https://nextjs.org/docs/app/api-reference/functions/generate-viewport
+export const viewport = DEFAULT_VIEWPORT;
+
+// This generates each page's HTML Metadata
+// @see https://nextjs.org/docs/app/api-reference/functions/generate-metadata
+export const generateMetadata = async (c: DynamicParams) => {
+  const { path = [], locale = defaultLocale.code } = c.params;
```
Comment (canerakdas) [inline app/[locale]/[[...path]]/page.tsx:23]:
```suggestion
export const generateMetadata = async ({params} : DynamicParams) => {
  const { path = [], locale = defaultLocale.code } = params;
```

`c` doesn't mean much, we can deconstruct directly

File: app/sitemap.ts [hunk 2]
```diff
@@ -1,37 +1,31 @@
...
 
-  // The current date of this request
   const currentDate = new Date().toISOString();
 
-  const appRoutes = [...dynamicRoutes, ...staticPaths]
-    .sort()
-    .map(route => `${baseUrlAndPath}/${route}`);
+  const appRoutes = paths.sort().map(route => `${baseUrlAndPath}/${route}`);
 
   return [...appRoutes, ...EXTERNAL_LINKS_SITEMAP].map(route => ({
```
Comment (canerakdas) [inline app/sitemap.ts:30]:
```suggestion
  for (const locale of availableLocaleCodes) {
    const routesForLanguage = await dynamicRouter.getRoutesByLanguage(locale);

    paths.push(...routesForLanguage.map(route => `${baseUrlAndPath}/${locale}/${route}`));
  }

  const currentDate = new Date().toISOString();

  return [...paths, ...EXTERNAL_LINKS_SITEMAP].map(route => ({
```

We can directly pass `baseUrlAndPath` no need to sort and map again

File: client-context.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
+import { cache } from 'react';
+
+import type { ClientSharedServerContext } from './types';
```
Comment (canerakdas) [inline client-context.ts:3]:
```suggestion
import type { ClientSharedServerContext } from '@/types';
```

**Review by canerakdas — COMMENTED** [review]
I have only minor review notes ✨  In addition to these changes, it would be good to update the part of the collaborator guide that points to the location of the layouts (https://github.com/nodejs/nodejs.org/blob/main/COLLABORATOR_GUIDE.md?plain=1#L143-L145)

**Comment by ovflowd** [comment]
> I have only minor review notes ✨ In addition to these changes, it would be good to update the part of the collaborator guide that points to the location of the layouts ([`main`/COLLABORATOR_GUIDE.md?plain=1#L143-L145](https://github.com/nodejs/nodejs.org/blob/main/COLLABORATOR_GUIDE.md?plain=1&rgh-link-date=2023-11-07T19%3A28%3A23Z#L143-L145))

@canerakdas can you open a new PR with these changes? (All of them)

**Pull request merged**